<a href="https://colab.research.google.com/github/cortiz313/Machine-Learning-Class/blob/main/Christian_Ortiz_01_quickstart.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# A quickstart introduction

## An example ...
As part of a [conservation effort](http://burrowingowlconservation.org/sightings/), Ann would like to report sightings of Burrowing Owls as she is hiking. Unfortunately, Ann doesn't know what a Burrowing Owl looks like so she goes to the web to look at pictures. What she has then, is a set of images that are labeled. By examining these labeled images she is **training** herself to recognize Burrowing Owls. 

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/owls2.png)


More generally, we can call this set of labeled images used for training, the **labeled training dataset**. Let's dive into this idea of a labeled training dataset a bit more. Suppose Clara is given the task of distinguishing between pictures of telecaster style guitars and stratocaster. But not to worry, because her boss has given her thousands of pictures of guitars. When looking at a picture, the only thing Clara knows is that it is of either a stratocaster or a telecaster. For example, here are some pictures of stratocasters and telecasters. 

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/guitars2.png)


Again, when looking at a picture all she knows is that it is a picture of a stratocaster or telecaster but she doesn't know which of the two it is. How long will it take for Clara to learn how to distinguish between these two guitar styles? Would the time be significantly shorter if her boss gave her 10,000 pictures or 100,000? If this is all the information she gets, she will never learn. What she needs is a **labeled** dataset. When presented with a picture she needs to know whether it is a picture of a stratocaster or a telecaster. 


**Now back to Ann learning to recognize Burrowing Owls**

When Ann is learning to recognize Burrowing Owls from her labeled training set, she is developing a model of what features make it a Burrowing Owl. Once she is done learning she can be on a hike, see an animal, and classify it as a Burrowing Owl or something else. This is an **inference** process---based on the evidence of different features of the animal she can infer what it is. And to throw more jargon at you, this type of problem is called a **classification problem**. In classification problems the system is given the features of an object and it needs to classify that object. For example,

* The features might be the words of a Twitter post (i.e., *Everything Everywhere All At Once was a f-ing masterpiece. I can't emphasize how great this movie is, it's just that great.*) and based on those features the system classifies the post as positve, negative, or neutral.
* The features might be the pixels of an image and based on those pixels the system classifies the image into one of 1,000 categories (it's an image of an owl, a bicycle).



## machine learning
In machine learning, classification systems have a similar two step process. First is the training phase where the system uses a labeled training dataset to build a model. (We will learn about the architecture of these models and how they learn a bit later.) 

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/quickDiagram2.png)



> **Supervised vs. Unsupervised learning**. When a machine learning system trains on labeled data this is called supervised learning. When a system learns with unlabeled data this is called unsupervised learning. A very common example of unsupervised learning is clustering where we might give our system a million unlabeled pictures and ask it to divide the images into 10 groups. 

Once we have a trained model, we can use that model for inference---we can give it pictures and the system can classify them as burrowing owl or something else. Again, the two phases are:

1. training
2. inference

In this introduction we are going to ignore the training phase and learn a bit about inference. To do that we are going to use a pre-trained model. What *pre-trained model* simply means is that someone else designed the architecture of the model and trained it. 

### AlexNet

The pretrained model we will use is AlexNet. AlexNet was designed by Alex Krishevsky, Ilya Sutskever, and Geoffrey Hinton. In 2012, AlexNet won a competition where the competing systems had to classify images into one of 1,000 categories. AlexNet had a 15% error rate and the error rate of the second place winner was over 25%. Since then there are dozens of systems that perform better, but we will use AlexNet because of its historic significance. The pretrained AlexNet was trained on a labeled training dataset of over one million images.

## Let's get started

#### First, a note ...
The intent of this notebook is for you to learn a little bit about data mining and have a bit of fun. The idea is not for you to understand every line of code. That will come later.

**Note:**

**First let's set the runtime to GPU (Graphics Processing Unit) -- click on 'runtime' in the menu above, select 'Change runtime type' and pick 'GPU'.**

Let's check to see if we set the runtime to GPU:

In [ ]:
!nvidia-smi

You should see something like:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   64C    P8    10W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

Showing, for example, that we are using a Tesla T4 GPU

Now we will set up a variable to allow us to use the GPU.

In [2]:
import torch
if torch.cuda.is_available():  
  dev = "cuda:0" 
else:  
  dev = "cpu"  
device = torch.device(dev) 
dev

'cuda:0'

### 1. Install Pytorch Lightning.
First, let's install the Pytorch Lightning library on our virtual machine

In [ ]:
!pip install pytorch-lightning


The exclamation point (aka *bang*) at the beginning of the line instructs the system to interpret the rest of the line as a Unix command (something you might type in a Unix terminal).

For example

```!ls```

will list the contents of the current directory

`pip` is the **p**ackage **i**nstaller for **P**ython. As the name suggests, it install Python libraries (packages) that are not already present in the system. In this case,

In the case above, we are installing the `pytorch-lightning` library.


### 2. Import the computer vision library


In [4]:
from torchvision import models
import torch

Without the bang Jupyter interprets code lines in this notebook as a Python commands which

```from torchvision import models``` is.

`torchvision` is a library containing models and other components for computer vision. `torch` is the basic PyTorch deep learning library.

There are many pretrained models available to use. Let's take a look at the possibilities:

In [ ]:
dir(models)

That is a lot of pretrained models we can use!

### 3. Load  the pretrained AlexNet model.

In [ ]:
alexnet = models.alexnet(weights='AlexNet_Weights.DEFAULT')
# have model run on the GPU
alexnet.cuda()

And let's check to see if the model is using the GPU ...

In [6]:
next(alexnet.parameters()).is_cuda

True

Excellent! CUDA is the API for NVIDIA GPUs that allow us to do parallel programming.

Of course we could call this model anything we want:

```
alexnet_model = models.alexnet(pretrained=True)
myModel = models.alexnet(pretrained=True)
```

Now we have a pretrained model loaded into our system. We would like to use the model to classify our own images. 

## Inference

We would like to give AlexNet an image like:

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/poodle.jpg)

and have AlexNet classify it. First, let's download the image file from the web using the Unix command `curl`:

In [ ]:
!curl http://zacharski.org/files/courses/dmpics/poodle.jpg -o poodle.jpg

Next, let's load that image into Python using PIL (Python Imaging Library):

In [8]:
from PIL import Image
img = Image.open("poodle.jpg")

Now let's display that image:

In [ ]:
img

(If that doesn't display an image change the code to `img.show`)


The size of this particular image is 4032x3024 which is slightly over 12 million pixels. That is a lot of pixels! AlexNet was designed to work with an image size of 224x224. So we need to transform the original image to AlexNet specifications by using some methods from the `torchvision` library. 

First we will use `transforms.Resize(256)` which transforms the image so that the smaller dimension of the original image will be resized to 256.  ([PyTorch documentation](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html)). The resultant image will be 341x256.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/poodleSmall.jpg)


To get the image to the final 224x224 size we are going to use `transforms.CenterCrop(224)` which as the name suggests crops the image at the center to a 224x224 square. The result will look something like:

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/poodleCropped.jpg)



Then we will convert the image to an array using `ToTensor`. Since each pixel has values for red, green and blue (RGB), the resulting array will be 224x224x3. Finally we normalize the tensor using `transforms.Normalize` ([PyTorch Documentation](https://pytorch.org/vision/main/generated/torchvision.transforms.Normalize.html?highlight=normalize#torchvision.transforms.Normalize)). 



Here are those transformations put together:



In [9]:
from torchvision import transforms
transform = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229,0.224,0.225])
])

You may wonder where the numbers come from in 

```
        mean=[0.485, 0.456, 0.406],
        std=[0.229,0.224,0.225])
```
These are the mean and standard deviation of the RGB values for all the pixels in all the images in the ImageNet dataset.

Let's use this method we defined to transform the image:


In [10]:
img_t = transform(img)

batch_t = torch.unsqueeze(img_t, 0)

# put the tensor on the GPU
batch_t = batch_t.to(device)
batch_t.is_cuda

True

Image classification models typically classify an array of images at once rather than a single image. 

```
    batch_t = torch.unsqueeze(img_t, 0)
```
creates a tensor with one element `img_t` which itself is a tensor. In a sense, an array of image arrays.

Next,

```
batch_t = batch_t.to(device)
```
Puts that Tensor on the GPU and

```
batch_t.is_cuda
```
checks to make sure that that is the case.

Now we have prepared the image and are prepared to pass it to alexnet for inference.

### Model Inference

In PyTorch, models can be in two modes and we can toggle between them.

* `alexnet.eval()` puts the model in inference mode so it can make predictions.
* `alexnet.train()` puts the model in training mode.


Let's get the model ready for inference:

In [ ]:
alexnet.eval()

As you can see, `alexnet.eval()` displays a lot of information about the architecture of the model. We will learn about the architecture of deep learning models about midway through the course.

Now let's pass the tensor of our image to alexnet and get the output. Plus, let's examine the shape of the output.

In [12]:
out = alexnet(batch_t)
print(out.shape)

torch.Size([1, 1000])


As we see `out` is a one dimensional tensor with 1,000 different values. We get 1,000 values because Image_net contained 1,000 labels for the images. The larger the number the more likely the image is of that class. 

In [ ]:
out

Let's convert those values to probabilities:

In [ ]:
percentage = torch.nn.functional.softmax(out, dim=1)[0] * 100
percentage

As you can see, the image is not very likely to be one of the first 5 labels. What are the actual names of these labels. First let's download the label name file:


In [ ]:
!curl http://zacharski.org/files/courses/dmpics/imagenet_classes.txt -o imagenet_classes.txt


Let's load in those labels:

In [ ]:
with open('imagenet_classes.txt') as f:

  labels = [line.strip() for line in f.readlines()]
print(labels[:5])

The first five labels are of types of fish. It is good to know our model didn't think our image was of a fish. What does our model think?


In [ ]:
z_, index = torch.max(out, 1)
print(index)

Okay, the label at index 267 is the most likely. Let's print that out:

In [ ]:
print(labels[index[0]], percentage[index[0]].item())


Fortunately, alexnet correctly thinks the image is of a standard poodle with 38.65% likelihood. 

The ImageNet competition evaluated systems based on the top-5 error rate meaning the system was judged correct if the correct label was among the top 5 the system predicted.  Let's look at the top 5 our model predicted for this image.

In [ ]:
_, indices = torch.sort(out, descending=True)
[(labels[idx], percentage[idx].item()) for idx in indices[0][:5]]


Let's turn what we learned into a function and try predicting the class of other images.

In [20]:
import  requests

def predict(url):
    # first download the image from the web
    r = requests.get(url)
    with open('tmp.jpg', 'wb') as f:
        f.write(r.content)
    img = Image.open('tmp.jpg')
    img.show()
    img_t = transform(img)
    batch_t = torch.unsqueeze(img_t, 0)
    batch_t = batch_t.to(device)
    out = alexnet(batch_t)
    _, indices = torch.sort(out, descending=True)
    percentage = torch.nn.functional.softmax(out, dim=1)[0] * 100
    return([(labels[idx], percentage[idx].item()) for idx in indices[0][:5]])


Let's find out what the 1,000 labels are:

In [ ]:

line = ''
for i in range(len(labels)):
  line += '%-25s' %labels[i]
  if (i + 1) % 4 == 0:
    print(line)
    line = ''

Electric Guitar is one of the labels. Let's see if our model correctly identifies a picture of one:

In [ ]:
predict('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/Fender_Stratocaster.jpeg')


Our system is 99% sure that this image is an electric guitar.

While AlexNet doesn't know about burrowing owls it does know about great horned owls. Let's give it a picture of a burrowing owl and see what it does:

In [ ]:
predict('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/greyOwl.jpeg')

That is a reasonable response!

Let's try a cello.

In [ ]:
predict('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/cello12.jpeg')

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/torchdivide.png)


# <font color='#EE4C2C'>You Try ...</font> 
Ok, it is time for you to try out what you just learned.

## <font color='#EE4C2C'>1. Your own images</font> 
Try this out on three images of your own.

In [24]:
# Small Dog
predict('https://static.wixstatic.com/media/1556ee_d0d16fca7f7647d98b13925809ae4500~mv2.jpg/v1/crop/x_579,y_0,w_3449,h_3442/fill/w_256,h_256,al_c,q_80,usm_0.66_1.00_0.01,enc_auto/P6250744_JPG.jpg')

[('Siberian husky', 13.124492645263672),
 ('dingo', 12.321903228759766),
 ('wire-haired fox terrier', 11.739368438720703),
 ('Eskimo dog', 11.098215103149414),
 ('borzoi', 8.581761360168457)]

In [25]:
# Octopus
predict('https://m.media-amazon.com/images/I/81Jn3ynx-ZL._CR0,204,1224,1224_UX256.jpg')

[('starfish', 43.36981201171875),
 ('pretzel', 32.398773193359375),
 ('hook', 11.302626609802246),
 ('knot', 4.323400497436523),
 ('chain', 3.29124116897583)]

In [26]:
# Lion
predict('https://d29rinwu2hi5i3.cloudfront.net/article_media/be642a8c-a3f7-4389-905b-b7dc18db52fd/headline_lion_2.jpg')

[('lion', 99.8720703125),
 ('dhole', 0.045167431235313416),
 ('chow', 0.033557455986738205),
 ('coyote', 0.023083439096808434),
 ('timber wolf', 0.014799346216022968)]

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/PyDivideTwo.png)
## <font color='#EE4C2C'>2. Squeezenet</font> 

Let's try a different pretrained model, `squeezenet1_1`. Load the model, construct a function that will make predictions based on the model and try it out on the images above that we provided plus your three.</span>

In [ ]:
# Setting up the model
squeezeModel = models.squeezenet1_1(weights='SqueezeNet1_1_Weights.DEFAULT')

squeezeModel.cuda()

next(squeezeModel.parameters()).is_cuda

squeezeModel.eval()


In [42]:
# The function to predict with the SqueezeNet model (predict2)
import  requests

def predict2(url):
    
    r = requests.get(url)
    with open('tmp.jpg', 'wb') as f:
        f.write(r.content)
    img = Image.open('tmp.jpg')
    img.show()
    img_t = transform(img)
    batch_t = torch.unsqueeze(img_t, 0)
    batch_t = batch_t.to(device)
    out = squeezeModel(batch_t)
    _, indices = torch.sort(out, descending=True)
    percentage = torch.nn.functional.softmax(out, dim=1)[0] * 100
    return([(labels[idx], percentage[idx].item()) for idx in indices[0][:5]])

In [43]:
# Guitar (given) SqueezeNet
predict2('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/Fender_Stratocaster.jpeg')

[('electric guitar', 98.95701599121094),
 ('acoustic guitar', 1.034130334854126),
 ('banjo', 0.0031488786917179823),
 ('violin', 0.001232030219398439),
 ('hook', 0.0010491503635421395)]

In [44]:
# Guitar (given) AlexNet
predict('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/Fender_Stratocaster.jpeg')


[('electric guitar', 99.0139389038086),
 ('acoustic guitar', 0.9524719715118408),
 ('rule', 0.019165046513080597),
 ('banjo', 0.0072016301564872265),
 ('hatchet', 0.0007767516653984785)]

In [31]:
# Owl (given) SqueezeNet
predict2('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/greyOwl.jpeg')

[('great grey owl', 82.93092346191406),
 ('ringlet', 10.850360870361328),
 ('puffer', 2.5810976028442383),
 ('African chameleon', 1.121286153793335),
 ('indri', 0.5488015413284302)]

In [32]:
# Owl (given) AlexNet
predict('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/greyOwl.jpeg')

[('great grey owl', 72.3724136352539),
 ('hare', 20.254365921020508),
 ('kite', 4.006711959838867),
 ('bittern', 1.0817829370498657),
 ('macaw', 0.6460889577865601)]

In [33]:
# Cello (given) SqueezeNet
predict2('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/cello12.jpeg')

[('cello', 65.8322982788086),
 ('violin', 34.16702651977539),
 ('acoustic guitar', 0.000545207760296762),
 ('hook', 6.222772935871035e-05),
 ('electric guitar', 3.828113040071912e-05)]

In [34]:
# Cello (given) AlexNet
predict('https://raw.githubusercontent.com/zacharski/ml-class/master/labs/pics/cello12.jpeg')

[('cello', 75.61215209960938),
 ('violin', 24.38234519958496),
 ('acoustic guitar', 0.0035001984797418118),
 ('banjo', 0.0019424731144681573),
 ('punching bag', 1.923392483149655e-05)]

In [35]:
# Small Dog SqueezeNet
predict2('https://static.wixstatic.com/media/1556ee_d0d16fca7f7647d98b13925809ae4500~mv2.jpg/v1/crop/x_579,y_0,w_3449,h_3442/fill/w_256,h_256,al_c,q_80,usm_0.66_1.00_0.01,enc_auto/P6250744_JPG.jpg')

[('wire-haired fox terrier', 56.12382125854492),
 ('Lakeland terrier', 7.058378219604492),
 ('tennis ball', 5.255655765533447),
 ('Irish terrier', 5.170267581939697),
 ('borzoi', 3.208376407623291)]

In [36]:
# Small Dog AlexNet
predict('https://static.wixstatic.com/media/1556ee_d0d16fca7f7647d98b13925809ae4500~mv2.jpg/v1/crop/x_579,y_0,w_3449,h_3442/fill/w_256,h_256,al_c,q_80,usm_0.66_1.00_0.01,enc_auto/P6250744_JPG.jpg')

[('Siberian husky', 13.124492645263672),
 ('dingo', 12.321903228759766),
 ('wire-haired fox terrier', 11.739368438720703),
 ('Eskimo dog', 11.098215103149414),
 ('borzoi', 8.581761360168457)]

In [37]:
# Octopus SqueezeNet
predict2('https://m.media-amazon.com/images/I/81Jn3ynx-ZL._CR0,204,1224,1224_UX256.jpg')

[('starfish', 89.46566009521484),
 ('knot', 1.8542699813842773),
 ('chain', 1.3044801950454712),
 ('pretzel', 1.156561017036438),
 ('night snake', 1.1037392616271973)]

In [38]:
# Octopus AlexNet
predict('https://m.media-amazon.com/images/I/81Jn3ynx-ZL._CR0,204,1224,1224_UX256.jpg')

[('starfish', 43.36981201171875),
 ('pretzel', 32.398773193359375),
 ('hook', 11.302626609802246),
 ('knot', 4.323400497436523),
 ('chain', 3.29124116897583)]

In [39]:
# Lion SqueezeNet
predict2('https://d29rinwu2hi5i3.cloudfront.net/article_media/be642a8c-a3f7-4389-905b-b7dc18db52fd/headline_lion_2.jpg')

[('lion', 99.99097442626953),
 ('chow', 0.005395764485001564),
 ('red wolf', 0.002386545529589057),
 ('dhole', 0.0008315872983075678),
 ('coyote', 0.00019514700397849083)]

In [40]:
# Lion AlexNet
predict('https://d29rinwu2hi5i3.cloudfront.net/article_media/be642a8c-a3f7-4389-905b-b7dc18db52fd/headline_lion_2.jpg')

[('lion', 99.8720703125),
 ('dhole', 0.045167431235313416),
 ('chow', 0.033557455986738205),
 ('coyote', 0.023083439096808434),
 ('timber wolf', 0.014799346216022968)]

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/PyDivideTwo.png)
## <font color='#EE4C2C'>2. Summary</font> 

Please answer the following questions by editing this markdown cell. </span>

1. Classification machine learning models have two modes. What are they?  
<font color='#EE4C2C'> Models can be in eval() mode, which prepares it to evaluate and predict what an image is, and train() mode, which is used to train the model. </font> 
2. What is a pretrained model? <font color='#EE4C2C'>A pretrained model is a model that has already been fed data and has been through the training process. This model is prepared to make an inference and a prediction on a new set of data.</font> 
2. What is supervised learning? What is unsupervised learning? 
<font color='#EE4C2C'> Supervised learning is when a machine learning system trains using labeled data. Unsupervised learning is when a system trains using unlabeled data. </font> 
3. Describe in a few sentences what Squeezenet is. (requires a bit of googling)
<font color='#EE4C2C'> Squeezenet is a deep neural network for computer vision that was released in 2016 from researchers at DeepScale, University of California Berkeley, and Stanford. The goal of the creators of SqueezeNet was to create a smaller neural network with equivalent accuracy to the larger ones with fewer parameters so that it can more appropriately fit into computer memory and can more easily be transmitted over a computer network. Upon release, SqueezeNet was described as having "AlexNet-level accuracy with 50x fewer parameters and < 0.5MB model size." Where AlexNet has 240MB of parameters, SqueezeNet has only 5MB. However, it is important to note that SqueezeNet is not just a 'squeezed' version of AlexNet, as they have entirely different architectures from one another.   </font> 
4. Compare the performance of AlexNet over Squeezenet. Was one more accurate than the other? Did you notice any other differences? 
<font color='#EE4C2C'> Overall, in my tests I would have to say that I found SqueezeNet to be more accurate for the data it was given. To be fair, in a few cases AlexNet was better. I found AlexNet to be about 1% more accurate at predicting the electric guitar, and 10% more accurate at predicting the cello. SqueezeNet was 10% more confident in its owl prediction, although as you mentioned in the lecture it is not exactly the same type of owl from their labels. However, the real differences came when I performed the test on the pictures I found myself. For my small dog picture, SqueezeNet predicted that it was a Wire-Haired Fox Terrier with 56% confidence, whereas AlexNet predicted that it was a Siberian Husky with 13% confidence. If you look at the picture, you will see that it is very likely to be a terrier of some kind, and looks nothing like a Husky. AlexNet did give it an 11% chance of being a terrier too, but again the SqueezeNet was much more accurate. For my Octopus picture, SqueezeNet was also more accurate, although neither had the appropriate label. SqueezeNet was 89% confident it was a Starfish, and AlexNet was 43% confident of the same thing, however AlexNet also gave it a 32% chance of being a pretzel, which made me laugh. For my final picture, both models were extremely accurate and had a nearly identical 99.9% confidence that it was a lion, which it was. If we take into account the reduced model size and how much fewer parameters SqueezeNet requires, it would seem like SqueezeNet is a much better choice for predicting the type of data that we did in these experiments. </font> 
